src/Builder/BuilderEncoder.php

```diff
@@ -0,0 +1,181 @@
+<?php
+
+namespace MongoDB\Builder;
+
+use LogicException;
+use MongoDB\Builder\Expression\ExpressionInterface;
+use MongoDB\Builder\Expression\FieldPath;
+use MongoDB\Builder\Expression\Variable;
+use MongoDB\Builder\Query\OrQuery;
+use MongoDB\Builder\Stage\GroupStage;
+use MongoDB\Builder\Stage\ProjectStage;
+use MongoDB\Builder\Stage\StageInterface;
+use MongoDB\Codec\EncodeIfSupported;
+use MongoDB\Codec\Encoder;
+use MongoDB\Exception\UnsupportedValueException;
+use stdClass;
+
+use function array_is_list;
+use function array_merge;
+use function get_object_vars;
+use function is_array;
+use function sprintf;
+
+/** @template-implements Encoder<Pipeline|StageInterface|ExpressionInterface, stdClass|array|string> */
+class BuilderEncoder implements Encoder
+{
+    use EncodeIfSupported;
+
+    /** The first property is the operator value */
+    public const ENCODE_AS_SINGLE = 'single';
+
+    /** Arguments as encoded as a map of properties to their values */
+    public const ENCODE_AS_OBJECT = 'object';
+
+    /** Properties are encoded as a list of values, names are ignored */
+    public const ENCODE_AS_ARRAY = 'array';
```

Comment:
Would an enum make more sense here? If the feature is going to require PHP 8.1+, I don't think there'd be any syntax limitation.